src/bin/pg_autoctl/supervisor.c

```diff
@@ -0,0 +1,827 @@
+							  services[idx].name, services[idx].pid);
+				}
+			}
+
+			/* we return false always, even if supervisor_stop is successful */
+			(void) supervisor_stop(&supervisor);
+
+			return false;
+		}
+	}
+
+	/*
+	 * We need to update our pid file with the PID for every service.
+	 */
+	if (!supervisor_update_pidfile(&supervisor))
+	{
+		log_fatal("Failed to update pidfile \"%s\", stopping all services now",
+				  supervisor.pidfile);
+
+		supervisor.cleanExit = false;
+		supervisor.shutdownSequenceInProgress = true;
+
+		(void) supervisor_stop_subprocesses(&supervisor);
+
+		return false;
+	}
+
+	/* now supervise sub-processes and implement retry strategy */
+	if (!supervisor_loop(&supervisor))
+	{
+		log_fatal("Something went wrong in sub-process supervision, "
+				  "stopping now. See above for details.");
+		success = false;
+	}
+
+	return supervisor_stop(&supervisor) && success;
+}
+
+
+/*
+ * service_supervisor calls waitpid() in a loop until the sub processes that
+ * implement our main activities have stopped, and then it cleans-up the PID
+ * file.
+ */
+static bool
+supervisor_loop(Supervisor *supervisor)
+{
+	int subprocessCount = supervisor->serviceCount;
+	int stoppingLoopCounter = 0;
+
+	/* wait until all subprocesses are done */
+	while (subprocessCount > 0)
+	{
+		pid_t pid;
+		int status;
+
+		/* Check that we still own our PID file, or quit now */
+		(void) check_pidfile(supervisor->pidfile, supervisor->pid);
+
+		/* If necessary, now is a good time to reload services */
+		if (asked_to_reload)
+		{
+			log_info("pg_autoctl received a SIGHUP signal, "
+					 "reloading configuration");
+			(void) supervisor_reload_services(supervisor);
+		}
+
+		/* ignore errors */
+		pid = waitpid(-1, &status, WNOHANG);
+
+		switch (pid)
+		{
+			case -1:
+			{
+				if (errno == ECHILD)
+				{
+					/* no more childrens */
+					if (asked_to_stop || asked_to_stop_fast)
+					{
+						/* off we go */
+						log_info("Internal subprocesses are done, stopping");
+						return true;
+					}
+				}
+				else
+				{
+					log_fatal("Failed to call waitpid(): %m");
+					return false;
+				}
+			}
+
+			case 0:
+			{
+				/*
+				 * We're using WNOHANG, 0 means there are no stopped or exited
+				 * children, it's all good. It's the expected case when
+				 * everything is running smoothly, so enjoy and sleep for
+				 * awhile.
+				 */
+				if (asked_to_stop || asked_to_stop_fast)
+				{
+					if (!supervisor->shutdownSequenceInProgress)
+					{
+						log_info("pg_autoctl received signal %s, terminating",
+								 asked_to_stop_fast ? "SIGINT" : "SIGTERM");
+
+						supervisor->cleanExit = true;
+						supervisor->shutdownSequenceInProgress = true;
+					}
+
+					/*
+					 * Stop all the services.
+					 */
+					if (stoppingLoopCounter == 0)
+					{
+						(void) supervisor_stop_subprocesses(supervisor);
+					}
+
+					(void) supervisor_shutdown_sequence(stoppingLoopCounter++);
```

Comment:
Can we move the `++`  outside of this call for clarity if the increment happens before or after the call?

Reply:
Move this bit of code and I think it's not ok?